device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java

```diff
@@ -830,6 +835,22 @@ private void checkForExpiredMessages()
         }
     }
 
+    private void checkForOldMessages()
+    {
+        // Check if the "correlationCallbacks" map contains any correlation ID which has existed over the
+        // default correlation ID live time. If so, remove the old correlation IDs from the map. Otherwise,
+        // the size of map will grow endlessly which results in OutOfMemory eventually.
+        for (String correlationId : correlationCallbacks.keySet())
+        {
+            if (System.currentTimeMillis() - correlationStartTimeMillis.get(correlationId) >= DEFAULT_CORRELATION_ID_LIVE_TIME)
+            {
+                correlationCallbacks.remove(correlationId);
```

Comment:
I'm pretty sure you can't remove a key from a map while iterating over that map's key set. You'll probably need to move this outside the foreach loop.